packages/dhis2/src/Adaptor.js

```diff
@@ -634,31 +513,33 @@ export function upsert(
       // NOTE: that these parameters are all expanded by the `get`, `create`, and
       // `update` functions used inside this composed "upsert" function.
       console.log(`Preparing upsert via 'get' then 'create' OR 'update'...`);
-      promise = get(
-        resourceType,
-        query,
-        options
-      )(state).then(resp => {
-        const resources = resp.data[resourceType];
-        if (resources.length > 1) {
-          throw new RangeError(
-            `Cannot upsert on Non-unique attribute. The operation found more than one records for your request.`
-          );
-        } else if (resources.length <= 0) {
-          return create(resourceType, data, options)(state);
-        } else {
-          // Pick out the first (and only) resource in the array and grab its
-          // ID to be used in the subsequent `update` by the path determined
-          // by the `selectId(...)` function.
-          const path = resources[0][selectId(resourceType)];
-          return update(resourceType, path, data, options)(state);
-        }
-      });
+      promise = http
```

Comment:
ooh this is really annoying

Because we're using the `http.get()` operation directly, we have to feed state back into the call here.

it's an antipattern really. We certainly don't want to encourage it, and this is a big, brand-defining adaptor.

We don't _have_ to do this in this PR, but what's the cost in using like `util.request` instead of `http.request`? Of using a function rather than an operation?

Also: if `asbase64` is true, I think this stuff might be broken? Because response.data is a string and all these tests will fail :grimacing: